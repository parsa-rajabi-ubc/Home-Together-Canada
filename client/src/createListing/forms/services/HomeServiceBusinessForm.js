/**
 * @Author:     Jeff Hatton
 * @Created:    2021.02.10
 *
 * @Description: Form to create a listing in Home Share Facilitation & Supporting Services
 * and Home Sharing Businesses, Groups and Organizations categories
 *
 */

import React, {useEffect, useState} from 'react';
import PropTypes from "prop-types";
import {HOME_SERVICE_BUSINESS_TEXT as TEXT} from "./constants/ServiceListingText";
import TextArea from "../../../common/forms/TextArea";
import LargeTextArea from "../../../common/forms/LargeTextArea";
import SubmitButton from "../../../common/forms/SubmitButton";
import {SHORT_DESC_CHAR_COUNT} from "../../../common/constants/listingsConstants";
import {
    checkIfErrorsExistInMapping,
    validateInput,
} from "../../../registration/registrationUtils";
import {validatePositiveNumber} from "../../../common/utils/generalUtils";
import Tooltip from "../../../common/forms/Tooltip";
import {CREATE_LISTING_MEMBER_SHARE_HOME as ToolTipText} from "../../../common/constants/TooltipText";
import {BUSINESS_SERVICE_CATEGORIES} from "../../constants/serviceListingCategoriesText";
import MultiImageUpload from "../../../common/forms/MultiImageUpload";
import {DEFAULT_MAX_NUM_IMAGES} from "../../constants/createListingConfig";
import {LISTING_FIELD_LENGTHS} from "../../../common/constants/fieldLengths";

const HomeServiceBusinessForm = (props) => {
    const {
        onSubmit,
        category,
        listingExists = false,
        existingTitle,
        existingShortDescription,
        existingFullDescription,
        existingRateAndFees
    } = props;

    const [title, setTitle] = useState(existingTitle || undefined);
    const [shortDescription, setShortDescription] = useState(existingShortDescription || undefined);
    const [rateAndFees, setRateAndFees] = useState(existingRateAndFees || undefined);
    const [fullDescription, setFullDescription] = useState(existingFullDescription || undefined);
    const [pictures, setPictures] = useState([]);

    const [titleError, setTitleError] = useState(undefined);
    const [shortDescriptionError, setShortDescriptionError] = useState(undefined);
    const [fullDescriptionError, setFullDescriptionError] = useState(undefined);
    const [rateAndFeesError, setRateAndFeesError] = useState(undefined);

    useEffect(() => {
        title !== undefined && validateInput(title, setTitleError);
    }, [title]);
    useEffect(() => {
        shortDescription !== undefined && validateInput(shortDescription, setShortDescriptionError);
    }, [shortDescription]);
    useEffect(() => {
        fullDescription !== undefined && validateInput(fullDescription, setFullDescriptionError);
    }, [fullDescription]);
    useEffect(() => {
        rateAndFees !== undefined && validatePositiveNumber(rateAndFees, setRateAndFeesError);
    }, [rateAndFees]);

    function handleImageUpload(e) {
        setPictures([...e.target.files]);
    }

    const isFormValid = () => {

        const errors = {
            title: false,
            shortDes: false,
            fullDes: false,
            rateAndFees: false,
        }

        errors.title = validateInput(title, setTitleError);
        errors.shortDes = validateInput(shortDescription, setShortDescriptionError);
        errors.fullDes = validateInput(fullDescription, setFullDescriptionError);
        errors.ratesAndFees = validatePositiveNumber(rateAndFees, setRateAndFeesError);

        return !(checkIfErrorsExistInMapping(errors));
    }

    //function for input checks on submit
    function onCreateListing() {
        if (isFormValid()) {
            onSubmit({
                title,
                shortDescription,
                fullDescription,
                rateAndFees,
                pictures
            });
        }
    }

    return (
        <div>
            <div className="col-start-1 col-end-7 py-5 px-5 m-6 bg-white shadow-lg rounded-xl">
                <div className="grid grid-cols-2 gap-6">
                    <div className="col-span-3 sm:col-span-2">
                        <h1 className={"page-title mb-5"}>
                            {(category === BUSINESS_SERVICE_CATEGORIES.SHARED_HOME_SERVICES
                                ? TEXT.home_form_title
                                : TEXT.business_form_title)}
                        </h1>
                        <TextArea
                            className={`${titleError && "border-red-500"} input`}
                            label={TEXT.title}
                            labelClassName={"label"}
                            required={true}
                            onChange={(e) => setTitle(e.target.value)}
                            value={title || ''}
                            charLimit={LISTING_FIELD_LENGTHS.TITLE}
                        />

                        <div className={"grid grid-cols-9"}>
                            <section className={"col-start-1 col-end-5"}>
                                <TextArea
                                    className={`${shortDescriptionError && "border-red-500"} input`}
                                    label={TEXT.short_des + " (" + SHORT_DESC_CHAR_COUNT + " Characters)"}
                                    labelClassName={"label"}
                                    required={true}
                                    onChange={(e) => setShortDescription(e.target.value)}
                                    value={shortDescription || ''}
                                    charLimit={SHORT_DESC_CHAR_COUNT}
                                />
                            </section>
                            <section className={"col-start-6 col-end-10"}>
                                <TextArea
                                    className={`${rateAndFeesError && "border-red-500"} input`}
                                    label={TEXT.rates_and_fees}
                                    labelClassName={"label"}
                                    required={true}
                                    onChange={(e) => setRateAndFees(e.target.value)}
                                    value={rateAndFees || ''}
                                    charLimit={LISTING_FIELD_LENGTHS.RATES_AND_FEES}
                                />
                            </section>
                        </div>
                        <LargeTextArea
                            className={`${fullDescriptionError && "border-red-500"} input`}
                            rows={"6"}
                            label={TEXT.full_des}
                            labelClassName={"label"}
                            required={true}
                            onChange={(e) => setFullDescription(e.target.value)}
                            value={fullDescription || ''}
                            charLimit={LISTING_FIELD_LENGTHS.FULL_DESCRIPTION}
                        />

                        {!listingExists &&
                        <div>
                            <label className="label"> Photos </label>
                            <p>{ToolTipText.PHOTOS}</p>
                            <MultiImageUpload handleImageUpload={handleImageUpload} maxNumImages={DEFAULT_MAX_NUM_IMAGES}/>
                        </div>
                        }
                    </div>
                </div>
                <SubmitButton
                    className={"btn btn-green form-btn w-1/2"}
                    onClick={onCreateListing}
                />
            </div>
        </div>
    );
}
HomeServiceBusinessForm.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    category: PropTypes.string.isRequired,
    listingExists: PropTypes.bool,
    existingTitle: PropTypes.string,
    existingShortDescription: PropTypes.string,
    existingFullDescription: PropTypes.string,
    existingRateAndFees: PropTypes.string,
}

export default HomeServiceBusinessForm;
