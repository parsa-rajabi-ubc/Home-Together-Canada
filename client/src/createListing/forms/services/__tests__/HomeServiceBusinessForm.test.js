/**
 * @Author:     Jeff Hatton
 * @Created:    2021.2.11
 *
 * @Description: Home share business services form Test
 *
 */

import React from 'react';
import renderer from  'react-test-renderer'
import HomeServiceBusinessForm from "../HomeServiceBusinessForm";

describe('HomeServiceBusinessForm', () => {
    describe('Snapshot test', () => {
        it("should render correctly regardless of properties", () => {
            // given
            const props = {
                onSubmit: jest.fn()
            };
            //when
            const component = renderer.create(<HomeServiceBusinessForm {...props}/>).toJSON();
            //then
            expect(component).toMatchSnapshot();
        });
    })
})