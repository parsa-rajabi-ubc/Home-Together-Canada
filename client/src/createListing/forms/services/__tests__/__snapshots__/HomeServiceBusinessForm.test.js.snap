// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeServiceBusinessForm Snapshot test should render correctly regardless of properties 1`] = `
<div>
  <div
    className="col-start-1 col-end-7 py-5 px-5 m-6 bg-white shadow-lg rounded-xl"
  >
    <div
      className="grid grid-cols-2 gap-6"
    >
      <div
        className="col-span-3 sm:col-span-2"
      >
        <h1
          className="page-title mb-5"
        >
          Home Share Facilitation & Supporting Services
        </h1>
        <section>
          <label
            className="label"
          >
            Descriptive Title
          </label>
          <svg
            className="inline align-top text-red-600"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
              fill="currentColor"
            />
          </svg>
          <input
            className="false undefined input"
            disabled={false}
            maxLength={75}
            onChange={[Function]}
            type="text"
            value=""
          />
        </section>
        <div
          className="grid grid-cols-9"
        >
          <section
            className="col-start-1 col-end-5"
          >
            <section>
              <label
                className="label"
              >
                Short Description (50 Characters)
              </label>
              <svg
                className="inline align-top text-red-600"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
                  fill="currentColor"
                />
              </svg>
              <input
                className="false undefined input"
                disabled={false}
                maxLength={50}
                onChange={[Function]}
                type="text"
                value=""
              />
            </section>
          </section>
          <section
            className="col-start-6 col-end-10"
          >
            <section>
              <label
                className="label"
              >
                Rates and fees
              </label>
              <svg
                className="inline align-top text-red-600"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
                  fill="currentColor"
                />
              </svg>
              <input
                className="false undefined input"
                disabled={false}
                maxLength={255}
                onChange={[Function]}
                type="text"
                value=""
              />
            </section>
          </section>
        </div>
        <div>
          <label
            className="label"
          >
            Full Description
          </label>
          <svg
            className="inline align-top text-red-600"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
              fill="currentColor"
            />
          </svg>
          <textarea
            className="undefined undefined undefined input"
            cols="50"
            maxLength={2000}
            onChange={[Function]}
            rows="6"
            value=""
          />
        </div>
        <div>
          <label
            className="label"
          >
             Photos 
          </label>
          <p>
            Each listing can have a maximum 6 pictures. Please use SHIFT or CONTROL to select pictures at once. You can NOT upload pictures one at a time.
          </p>
          <div
            className="flex justify-center px-6 pt-5 pb-6 border-2 border-gray-300 border-dashed rounded-md"
          >
            <div
              className="space-y-1 text-center"
            >
              <svg
                aria-hidden="true"
                className="w-12 h-12 mx-auto text-gray-400"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 48 48"
              >
                <path
                  d="M28 8H12a4 4 0 00-4 4v20m32-12v8m0 0v8a4 4 0 01-4 4H12a4 4 0 01-4-4v-4m32-4l-3.172-3.172a4 4 0 00-5.656 0L28 28M8 32l9.172-9.172a4 4 0 015.656 0L28 28m0 0l4 4m4-24h8m-4-4v8m-12 4h.02"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                />
              </svg>
              <input
                accept="image/png, image/jpg, image/jpeg, image/JPG, image/JPEG"
                className="photo-upload-width photo-upload hover:text-indigo-500"
                multiple={true}
                name="logo"
                onChange={[Function]}
                type="file"
              />
              <p
                className="text-xs text-gray-500"
              >
                PNG or JPG up to 2MB
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <label>
      <input
        className="btn btn-green form-btn w-1/2"
        disabled={false}
        onClick={[Function]}
        type="submit"
      />
    </label>
  </div>
</div>
`;
